File: Sources/AnagramEngine/Phonetics.swift
import Foundation

/// Phonetic queries backed by the CMU Pronouncing Dictionary.
///
/// CMUdict line format (cmudict.dict):  `word P H ON EM`
/// Vowel phonemes carry a stress digit (0/1/2); consonants don't. Alternate
/// pronunciations appear as `word(2)`. Lines may end with a `#` comment.
public struct PhoneticDictionary: Sendable {
    /// word -> list of pronunciations (each a phoneme array)
    private let pronunciations: [String: [[String]]]
    /// perfect-rhyme key (last stressed vowel → end) -> words
    private let rhymeIndex: [String: Set<String>]
    /// full pronunciation, stress stripped -> words (for homophones)
    private let homophoneIndex: [String: Set<String>]

    public init(cmudictText text: String) {
        var pron: [String: [[String]]] = [:]
        var rhyme: [String: Set<String>] = [:]
        var homo: [String: Set<String>] = [:]

        for rawLine in text.split(separator: "\n") {
            // Strip trailing comment.
            var line = Substring(rawLine)
            if let hash = line.firstIndex(of: "#") { line = line[..<hash] }
            let tokens = line.split(whereSeparator: { $0 == " " || $0 == "\t" })
            guard tokens.count >= 2 else { continue }

            // Word, dropping any "(2)" variant suffix.
            var word = String(tokens[0]).lowercased()
            if let paren = word.firstIndex(of: "(") { word = String(word[..<paren]) }
            guard word.allSatisfy({ ($0 >= "a" && $0 <= "z") || $0 == "'" || $0 == "-" }) else { continue }

            let phonemes = tokens[1...].map(String.init)
            pron[word, default: []].append(phonemes)

            if let key = Self.rhymeKey(phonemes) {
                rhyme[key, default: []].insert(word)
            }
            let homoKey = phonemes.map(Self.stripStress).joined(separator: " ")
            homo[homoKey, default: []].insert(word)
        }

        pronunciations = pron
        rhymeIndex = rhyme
        homophoneIndex = homo
    }

    public var count: Int { pronunciations.count }
    public func isKnown(_ word: String) -> Bool { pronunciations[word.lowercased()] != nil }

    // MARK: Syllables

    /// Syllable count = number of vowel phonemes (those carrying a stress digit).
    /// Returns nil if the word isn't in the dictionary.
    public func syllableCount(_ word: String) -> Int? {
        guard let prons = pronunciations[word.lowercased()], let first = prons.first else { return nil }
        return first.filter { $0.contains(where: \.isNumber) }.count
    }

    // MARK: Rhymes

    /// Words that perfectly rhyme with `word` (share the sound from the last
    /// stressed vowel onward), excluding the word itself.
    public func rhymes(_ word: String) -> [String] {
        let w = word.lowercased()
        guard let prons = pronunciations[w] else { return [] }
        var result: Set<String> = []
        for p in prons {
            if let key = Self.rhymeKey(p) {
                result.formUnion(rhymeIndex[key] ?? [])
            }
        }
        result.remove(w)
        return result.sorted()
    }

    // MARK: Homophones

    /// Words pronounced identically to `word` (different spelling).
    public func homophones(_ word: String) -> [String] {
        let w = word.lowercased()
        guard let prons = pronunciations[w] else { return [] }
        var result: Set<String> = []
        for p in prons {
            let key = p.map(Self.stripStress).joined(separator: " ")
            result.formUnion(homophoneIndex[key] ?? [])
        }
        result.remove(w)
        return result.sorted()
    }

    // MARK: Raw access (for FusionFinder and friends)

    /// All pronunciations of `word` (raw phonemes, stress digits intact).
    public func pronunciations(of word: String) -> [[String]] {
        pronunciations[word.lowercased()] ?? []
    }

    /// Words whose full stress-stripped pronunciation equals `phones`.
    public func words(pronouncedAs phones: [String]) -> Set<String> {
        homophoneIndex[phones.joined(separator: " ")] ?? []
    }

    /// Every word with its first (primary) pronunciation, stress stripped —
    /// the raw material for phoneme-level neighborhood searches.
    public func allStrippedPronunciations() -> [(word: String, phones: [String])] {
        pronunciations.compactMap { word, prons in
            guard let first = prons.first else { return nil }
            return (word, first.map(Self.stripStress))
        }
    }

    // MARK: Helpers

    private static func stripStress(_ phoneme: String) -> String {
        String(phoneme.filter { !$0.isNumber })
    }

    /// Key for perfect rhyme: phonemes from the last stressed vowel to the end.
    /// Falls back to the last vowel if none carry primary/secondary stress.
    static func rhymeKey(_ phonemes: [String]) -> String? {
        var lastStressed = -1
        var lastVowel = -1
        for (i, p) in phonemes.enumerated() where p.contains(where: \.isNumber) {
            lastVowel = i
            if p.contains("1") || p.contains("2") { lastStressed = i }
        }
        let start = lastStressed >= 0 ? lastStressed : lastVowel
        guard start >= 0 else { return nil }
        return phonemes[start...].map(stripStress).joined(separator: " ")
    }
}

File: app/src/engine/phonetics.ts
// Ported from Sources/AnagramEngine/Phonetics.swift

/**
 * Phonetic queries backed by the CMU Pronouncing Dictionary.
 *
 * CMUdict line format (cmudict.dict):  `word P H ON EM`
 * Vowel phonemes carry a stress digit (0/1/2); consonants don't. Alternate
 * pronunciations appear as `word(2)`. Lines may end with a `#` comment.
 */
export class PhoneticDictionary {
  /** word -> list of pronunciations (each a phoneme array) */
  private readonly prons: Map<string, string[][]>;
  /** perfect-rhyme key (last stressed vowel → end) -> words */
  private readonly rhymeIndex: Map<string, Set<string>>;
  /** full pronunciation, stress stripped -> words (for homophones) */
  private readonly homophoneIndex: Map<string, Set<string>>;

  constructor(cmudictText: string) {
    this.prons = new Map();
    this.rhymeIndex = new Map();
    this.homophoneIndex = new Map();

    for (const rawLine of cmudictText.split("\n")) {
      // Strip trailing comment.
      let line = rawLine;
      const hash = line.indexOf("#");
      if (hash !== -1) line = line.slice(0, hash);
      const tokens = line.split(/[ \t]+/).filter((t) => t.length > 0);
      if (tokens.length < 2) continue;

      // Word, dropping any "(2)" variant suffix.
      let word = tokens[0]!.toLowerCase();
      const paren = word.indexOf("(");
      if (paren !== -1) word = word.slice(0, paren);
      if (!/^[a-z'-]+$/.test(word)) continue;

      const phonemes = tokens.slice(1);
      let list = this.prons.get(word);
      if (!list) this.prons.set(word, (list = []));
      list.push(phonemes);

      const key = PhoneticDictionary.rhymeKey(phonemes);
      if (key !== null) {
        let bucket = this.rhymeIndex.get(key);
        if (!bucket) this.rhymeIndex.set(key, (bucket = new Set()));
        bucket.add(word);
      }
      const homoKey = phonemes.map(stripStress).join(" ");
      let hb = this.homophoneIndex.get(homoKey);
      if (!hb) this.homophoneIndex.set(homoKey, (hb = new Set()));
      hb.add(word);
    }
  }

  /**
   * Fast path for the baked lexicon format (`word|PHONES|RHYMEKEY`, one line
   * per word, phones already stress-stripped, rhyme key precomputed — see
   * scripts/build-lexicon.mjs). Skips the cmudict parse entirely.
   */
  static fromLexicon(text: string): PhoneticDictionary {
    const dict = new PhoneticDictionary("");
    for (const line of text.split("\n")) {
      if (line.length === 0) continue;
      const [word, ph, rk] = line.split("|");
      if (!word || !ph) continue;
      const phones = ph.split(" ");
      let list = dict.prons.get(word);
      if (!list) dict.prons.set(word, (list = []));
      list.push(phones);
      if (rk) {
        let bucket = dict.rhymeIndex.get(rk);
        if (!bucket) dict.rhymeIndex.set(rk, (bucket = new Set()));
        bucket.add(word);
      }
      let hb = dict.homophoneIndex.get(ph);
      if (!hb) dict.homophoneIndex.set(ph, (hb = new Set()));
      hb.add(word);
    }
    return dict;
  }

  get count(): number {
    return this.prons.size;
  }

  isKnown(word: string): boolean {
    return this.prons.has(word.toLowerCase());
  }

  /**
   * Syllable count = number of vowel phonemes. Returns null if the word isn't
   * in the dictionary. (Swift counts stress digits; every ARPABET vowel starts
   * with A/E/I/O/U and no consonant does, so this test is equivalent on
   * cmudict input — and, unlike the digit test, also works on the baked
   * lexicon, whose phones are stress-stripped.)
   */
  syllableCount(word: string): number | null {
    const first = this.prons.get(word.toLowerCase())?.[0];
    if (!first) return null;
    return first.filter((p) => /^[AEIOU]/.test(p)).length;
  }

  /**
   * Words that perfectly rhyme with `word` (share the sound from the last
   * stressed vowel onward), excluding the word itself.
   */
  rhymes(word: string): string[] {
    const w = word.toLowerCase();
    const prons = this.prons.get(w);
    if (!prons) return [];
    const result = new Set<string>();
    for (const p of prons) {
      const key = PhoneticDictionary.rhymeKey(p);
      if (key !== null) {
        for (const r of this.rhymeIndex.get(key) ?? []) result.add(r);
      }
    }
    result.delete(w);
    return [...result].sort();
  }

  /** Words pronounced identically to `word` (different spelling). */
  homophones(word: string): string[] {
    const w = word.toLowerCase();
    const prons = this.prons.get(w);
    if (!prons) return [];
    const result = new Set<string>();
    for (const p of prons) {
      const key = p.map(stripStress).join(" ");
      for (const h of this.homophoneIndex.get(key) ?? []) result.add(h);
    }
    result.delete(w);
    return [...result].sort();
  }

  /** All pronunciations of `word` (raw phonemes, stress digits intact). */
  pronunciations(word: string): string[][] {
    return this.prons.get(word.toLowerCase()) ?? [];
  }

  /** Words whose full stress-stripped pronunciation equals `phones`. */
  wordsPronouncedAs(phones: string[]): Set<string> {
    return this.homophoneIndex.get(phones.join(" ")) ?? new Set();
  }

  /**
   * Every word with its first (primary) pronunciation, stress stripped —
   * the raw material for phoneme-level neighborhood searches.
   */
  allStrippedPronunciations(): Array<{ word: string; phones: string[] }> {
    const out: Array<{ word: string; phones: string[] }> = [];
    for (const [word, prons] of this.prons) {
      const first = prons[0];
      if (first) out.push({ word, phones: first.map(stripStress) });
    }
    return out;
  }

  /**
   * Key for perfect rhyme: phonemes from the last stressed vowel to the end.
   * Falls back to the last vowel if none carry primary/secondary stress.
   */
  static rhymeKey(phonemes: string[]): string | null {
    let lastStressed = -1;
    let lastVowel = -1;
    for (let i = 0; i < phonemes.length; i++) {
      const p = phonemes[i]!;
      if (/\d/.test(p)) {
        lastVowel = i;
        if (p.includes("1") || p.includes("2")) lastStressed = i;
      }
    }
    const start = lastStressed >= 0 ? lastStressed : lastVowel;
    if (start < 0) return null;
    return phonemes.slice(start).map(stripStress).join(" ");
  }
}

/** Remove the stress digit from a phoneme ("AE1" → "AE"). */
export function stripStress(phoneme: string): string {
  return phoneme.replace(/\d/g, "");
}
